### PR TITLE
fix #1509 Added has discarded matching method in StepVerifier

### LINKED ISSUE #1509: Missing hasDiscardedMatching in StepVerifier
Right now doOnDiscard test for element equality, we would need a version to introspect the data (possibly with optional type argument to cast Object).

### PR DESCRIPTION
**Pull request by mumukiller**
Fix #1509 
Missed hasDiscardedMatching in StepVerifier was added

### CODE REVIEW

**Comment by mumukiller** [comment]
That's weird! I run FluxDistinctUntilChangedTest and it passed. Any ideas what's wrong with this test?

**Comment by simonbasle** [comment]
@mumukiller I think you can look at `hasDroppedErrorMatching`, `hasDroppedErrorsMatching` and `hasDroppedErrorsSatisfying` for inspiration of a feature-complete assertion of discarded elements.

**Comment by mumukiller** [comment]
@simonbasle do you suggest to make hasDiscardedMatching behaviour the same as hasDroppedErrorMatching?

These three - hasDroppedErrorMatching, hasDroppedErrorsMatching and hasDroppedErrorsSatisfying are about exceptions not about discarded elements.

**Comment by simonbasle** [comment]
Yeah, if we are to introduce `hasDiscardedMatching` it should have the same type of behavior as other `hasXxx_singularMatching`, namely check there is only one element discarded AND it matches the predicate.

But I think this was probably a mistake to have that single-element checking method for dropped errors in the first place. Let's focus on matching/asserting the **collection** of discarded elements, with method names that leave less room for ambiguity:

 - `hasDiscardedElementsMatching(Predicate<Collection<Object>>)`
 - `hasDiscardedElementsSatisfying(Consumer<Collection<Object>>)` (this is actually the most general case, capable of covering every other when using AssertJ)

File: reactor-test/src/main/java/reactor/test/StepVerifier.java [hunk 1]
```diff
@@ -1072,10 +1072,17 @@ static void resetDefaultTimeout() {
 		Assertions hasDiscardedExactly(Object... values);
 
 		/**
-		 * Assert that the tested publisher has discarded at least one element matching the given
-		 * predicate to the {@link reactor.core.publisher.Flux#doOnDiscard(Class, Consumer) discard} hook.
+		 * Assert that the tested publisher has discarded one or more errors to the
```
Comment (simonbasle) [inline reactor-test/src/main/java/reactor/test/StepVerifier.java]: should be "elements", not "errors".

File: reactor-test/src/main/java/reactor/test/StepVerifier.java [hunk 2]
```diff
@@ -1072,10 +1072,17 @@ static void resetDefaultTimeout() {
...
+		 * {@link reactor.core.publisher.Flux#doOnDiscard(Class, Consumer) discard} hook,
+		 * and check that the collection of discarded elements matches a predicate.
 		 */
-		Assertions hasDiscardedMatching(Predicate<Object> matcher);
+		Assertions hasDiscardedElementsMatching(Predicate<Collection<Object>> matcher);
+
+		/**
+		 * Assert that the tested publisher has discarded one or more elements matching to the
```
Comment (simonbasle) [inline reactor-test/src/main/java/reactor/test/StepVerifier.java]: remove the "matching" at the end of the line

File: reactor-test/src/main/java/reactor/test/StepVerifier.java (hunk 1 above)
Comment (mumukiller) [reply reactor-test/src/main/java/reactor/test/StepVerifier.java]: done

File: reactor-test/src/main/java/reactor/test/StepVerifier.java (hunk 2 above)
Comment (mumukiller) [reply reactor-test/src/main/java/reactor/test/StepVerifier.java]: done

**Comment by simonbasle** [comment]
Thanks for the added feature @mumukiller 🙇🎁

**Pull request merged**